Implement ACL

erw7 opened the pull request:
Continue to #4060.

heewa committed:
acl: move fns and types to os/ & remove guards

OS-specific implementation to come in a separate commit.

heewa committed:
acl: implement sys/acl.h based backend

erw7 committed:
Change to detect libacl

erw7 committed:
Change to remove execute permission from ACL of undo file

erw7 committed:
Fix lint issue

erw7 committed:
Add ACL implementation on Windows

erw7 committed:
Fix compilation warnings

erw7 committed:
Add libacl1-dev and libacl1-dev:i386

erw7 committed:
Fix header file name

justinmk commented:
Leak is being reported in various tests (always from `<unknown module>`):

```
= =================================================================
= ==16110==ERROR: LeakSanitizer: detected memory leaks
= 
= Direct leak of 142 byte(s) in 6 object(s) allocated from:
=     #0 0x50b068 in malloc (/home/travis/build/neovim/neovim/build/bin/nvim+0x50b068)
=     #1 0x1e14449 in uv__strdup /home/travis/nvim-deps/build/src/libuv/src/uv-common.c:57
=     #2 0x65206d6574737972  (<unknown module>)
= 
= SUMMARY: AddressSanitizer: 142 byte(s) leaked in 6 allocation(s).
=====================================================================================================
test/helpers.lua:152: assertion failed!
stack traceback:
	test/helpers.lua:152: in function 'check_logs'
```

erw7 committed:
Change condition of HAVE_ACL

erw7 committed:
Change to use strerror instead of uv_strerror

Do not use of uv_strerror because a "Unknow system error" causes a
memory leak.

erw7 committed:
Fix check-single-includes error [skip appveyor]

erw7 commented:
https://travis-ci.org/neovim/neovim/jobs/552662869#L1656-L1658

```
ERROR 2019-07-01T08:46:23.866 11868 u_write_undo:1111: failed set permset to entry: No data available
ERROR 2019-07-01T08:46:23.866 11868 u_write_undo:1111: failed set permset to entry: Success
ERROR 2019-07-01T08:46:23.866 11868 u_write_undo:1111: failed set permset to entry: Success
```

According to `acl_set_permset (3)`, it is impossible .

erw7 commented:
https://travis-ci.org/neovim/neovim/jobs/552662869#L1654-L1655

```
ERROR 2019-07-01T08:46:23.662 11865 os_get_acl:94: failed to get acl of a file: No such file or directory
ERROR 2019-07-01T08:46:23.732 11866 os_get_acl:94: failed to get acl of a file: No such file or directory
```

Is there a route for this to occur? Should I check the existence of the file?

erw7 committed:
Change to disable libacl on OpenBSD, NetBSD

erw7 committed:
Fix build failure on Windows

justinmk commented on .travis.yml:
```diff
@@ -37,6 +37,10 @@ env:
+    # Additional CMake flags for 32-bit builds.
+    - CMAKE_FLAGS_32BIT="-DCMAKE_SYSTEM_LIBRARY_PATH=/lib32:/usr/lib32:/usr/local/lib32
+                         -DCMAKE_IGNORE_PATH=/lib:/usr/lib:/usr/local/lib
+                         -DCMAKE_TOOLCHAIN_FILE=$TRAVIS_BUILD_DIR/cmake/i386-linux-gnu.toolchain.cmake"
```
why is this needed?

erw7 replied:
I don't know what the cause is, but in the CI environment 64bit `libacl` was used by linking and the build failed, so I needed to set it.

justinmk replied:
This should be needed. We used to have something like this, but removed it recently: 3d3c783fead3dc328ede3ccd0fb7b3befcad191a

maybe @blueyed has insight

justinmk commented on CMakeLists.txt:
```diff
@@ -98,6 +98,13 @@ else()
+  option(ENABLE_LIBACL "enable libacl" OFF)
+else()
+  option(ENABLE_LIBACL "enabled libacl" ON)
+endif()
```
does this still need to be a cmake option? why not let `find_package(LibACL)` decide it?

erw7 replied:
I just followed it because `libintl` and `libiconv` are cmake option without automatic detection.

justinmk replied:
ok. `libintl` and `libiconv` are that way because of history (Windows support). We will remove those options once we're confident that they are always working on Windows.

In the case of ACL, there's no reason to make it an option: either we find the library or we don't.

erw7 committed:
Change to auto detect the presence of libacl

janlazo commented:
@justinmk @erw7 Is this ready?

erw7 commented:
Sorry for the delay in replying. I've been using neovim with this patch applied on a regular basis and haven't had any problems, so I don't think it's a problem to merge it.

tombh commented:
What's the status of this? I don't believe ACLs are currently supported in Neovim?

In fact I don't even think traditional `umask` and `chmod g+s ...` are currently supported? For instance [new](https://github.com/neovim/neovim/blob/6696ea7f103814d3d5700107546280bf50a4004a/src/nvim/os/fileio.c#L78) [path](https://github.com/neovim/neovim/blob/6696ea7f103814d3d5700107546280bf50a4004a/src/nvim/ex_cmds.c#L1737) creations simply harcode `755` permissions.

justinmk commented:
status is:

- https://github.com/neovim/neovim/issues/1200
- https://github.com/neovim/neovim/issues/26127#issuecomment-1829623039
